PLT-7111 Add groupmsg command

jespino opened the pull request:
Ticket Link
-----------

mattermost/mattermost-server#6932

Checklist
---------

- [X] Added or updated unit tests (required for all new features)
- [X] Includes text changes and localization file

jespino commented:
I want to explain a bit the PR:

1. I'm not a Go expert, actually I haven't written to much Go code, so the code can have some pretty obvious mistakes.
2. I based my code in command_msg.go.
3. I created it in 3 commits to simplify the revision process (I can rebase it to a one commit when you want). The first commit directly copy command_msg.go to command_groupmsg.go, the second one, change msg/MSG/Msg in the file to groupmsg/GROUPMSG/Groupmsg, and finally the third command is which do the work. So is easier to review the last one directly.
4. I based the tests on the command_msg_test.go code too, so Is very similar too. I'm not sure if is needed more testing. I think can be good to, have all error cases covered. Probably tomorrow I will try to cover it. (Using the same policy of command_msg, which is run simple requests to fire error cases, but without checking the error response, If you prefer, I can check the error response).

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

jespino commented:
Here is a weird case. If I include myself in the groupmsg and other guy, is a 2 people conversation, so this must fail (because it isn't a group message). Then the error shown is "Group messages are limited to a minimun of 2 users.". But I don't see it very intuitive, maybe the errors must be something like "Group messages are limited to a minimun of 2 users and you." y "Group messages are limited to a maxium of 7 users and you."

mattermod commented:
Spinmint test server created at: http://i-0c903ab5495ff10e2.spinmint.com:8065/pr7419

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0c903ab5495ff10e2

esethna commented:
Thanks for the PR @jespino, apologies it's taking a while to review given that the test servers are having issues after the repo split. 

Regarding your comment about using the /groupmsg command when only one user is listed. I think in order to follow our principle of [forgiving](https://docs.mattermost.com/developer/fx-guidelines.html?highlight=forgiving#fast-obvious-forgiving), we should just open up the direct message channel with the user, rather than show a system error message.

jespino commented:
Perfect, I couldn't be more agree.

jespino commented:
Fixed (and rebased to master the branch). I added a new commit using the approach of fallback to msg command on groupmsg with only one other user. I'm spliting it in different commits to easy the review process, when the branch is ready to merge I can flat the commits in a single one. :)

mattermod commented:
Spinmint test server created at: http://i-019127d5807fddeb2.spinmint.com:8065/pr7419

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-019127d5807fddeb2

jespino commented:
I've noticed that I have to manage duplications.

jespino commented:
Now added duplication management.

mattermod commented:
Spinmint test server destroyed

hmhealey commented:
Sorry @jespino, I was deleting mattermod comments so that I could update the test server and accidentally deleted one of yours

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-083d154e7e51b9ba5.spinmint.com:8065/pr7419

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-083d154e7e51b9ba5

jespino commented:
I think the tag "Awaiting Submitter Action" is no longer needed.

jespino commented:
Rebased

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-009b4c1b5b157bd97.spinmint.com:8065/pr7419

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-009b4c1b5b157bd97

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-04428889cc8316a5a.spinmint.com:8065/pr7419

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-04428889cc8316a5a

mattermod commented:
Spinmint test server destroyed

hmhealey commented on app/command_groupmsg.go:
```diff
@@ -0,0 +1,154 @@
+
+	if len(targetUsersSlice) < model.CHANNEL_GROUP_MIN_USERS {
+		return &model.CommandResponse{
+			Text:         fmt.Sprintf(args.T("api.command_groupmsg.min_users.app_error"), model.CHANNEL_GROUP_MIN_USERS-1),
```
Similarly with the rest of these translations, you don't need a fmt.Sprintf if you define the translation string like
```
"translation": "Group messages are limited to a minimun of {{.MinUsers}} users."
```
and call it with
```go
args.T("api.command_groupmsg.min_users.app_error", map[string]interface{"MinUsers": ...})
```

hmhealey commented on app/command_groupmsg.go:
```diff
@@ -0,0 +1,154 @@
+	} else {
+		channel := channel.Data.(*model.Channel)
+		targetChannelId = channel.Id
+		teamId = channel.TeamId
```
All group message channels have an empty teamId, so this will always be empty. That'll simplify the logic below a bit since you won't need to check the session, and you can just use `args.TeamId` when getting the team name to construct the URL.

That'll also prevent it from changing teams since I think it may do that right now when the current team doesn't match up with the one for `args.Session.TeamMembers[0]`

jespino commented:
I have applied the changes proposed in a new commit. Remember that if you want I can rebase it to a unique commit or split however you want (something like, copy from command_msg and then the rest of the changes).

jespino commented:
I pushed the changes to my own branch, but is not showed here :(

ccbrown commented on app/command_groupmsg.go:
```diff
@@ -0,0 +1,148 @@
+	channelName := model.GetGroupNameFromUserIds(targetUsersSlice)
+
+	targetChannelId := ""
+	if channel := <-a.Srv.Store.Channel().GetByName(args.TeamId, channelName, true); channel.Err != nil {
```
I'm pretty sure `(*App).CreateGroupChannel` will return the existing channel if there already is one. If so, you can simplify this a lot.

jespino replied:
Yes, you are right, I changed it and simplified the code. :)

jespino replied:
I removed the regexp and change with the code and some unit tests to ensure that is working correctly :)

jespino replied:
And I fixed the "4 spaces" style, I have to configure correctly my vim to reformat it correctly.

ccbrown commented on api/command_groupmsg_test.go:
```diff
@@ -0,0 +1,59 @@
+import (
+	"strings"
+	"testing"
+    "app"
```
```
api/command_groupmsg_test.go:9:5: cannot find package "app" in any of:
	/home/ubuntu/workspace/mattermost-platform-pr-split/src/github.com/mattermost/mattermost-server/vendor/app (vendor tree)
	/usr/local/go/src/app (from $GOROOT)
	/home/ubuntu/workspace/mattermost-platform-pr-split/src/app (from $GOPATH)
```

jespino replied:
Fixed. Thanks for your patience.

hmhealey commented:
@jespino Looks of, but it needs to be rebased or have master merged into it. And you don't need to worry about squashing it since GitHub does that automatically

jespino commented:
Ready to merge. :smiley:

jwilander commented:
Awesome 👍  I'll merge after we cut the release branch on Monday

jwilander commented:
@jespino mind giving this a rebase? The build doesn't like it when a PR gets too far behind from the latest master commits

jespino commented:
Done! :smile:

jwilander commented:
Sorry for the late merge, somehow flew under my radar

jespino commented:
No problem, and thanks! :smiley:
